### PR TITLE
[pkg/errors] Add Gatherer DoneOrFailed(Ctx)

### LINKED ISSUE #181: pkg/errors.Gatherer: Add `DoneOrFailed()` or similar.
Currently, it's impossible to perform a "Wait for completion of all goroutines or until the first goroutine fails" trivially. We need that behavior in the `erdstall` project. Please either add a function `DoneOrFailed() <-chan error` (or `<-chan struct{}`, I don't really care), or a `WaitDoneOrFailed()` function. While you're at it, it might also be an appropriate time to add context support for the Gatherer, but that's optional.

### PR DESCRIPTION
**Pull request by ggwpez**
Closes #181  
Need this for testing, @RmbRT also seems to appreciate it.

### CODE REVIEW

File: pkg/errors/gatherer.go [hunk 1]
```diff
@@ -44,6 +47,32 @@ func (g *Gatherer) Failed() <-chan struct{} {
...
+// any routine returned with an error.
+func (g *Gatherer) DoneOrFailed() <-chan struct{} {
+	return g.DoneOrFailedCtx(context.Background())
+}
+
+// DoneOrFailedCtx closes the returned channel when either:
+// all routines returned successfully OR
+// any routine returned with an error OR
+// the context was cancelled.
+func (g *Gatherer) DoneOrFailedCtx(ctx context.Context) <-chan struct{} {
```
Comment (RmbRT) [inline pkg/errors/gatherer.go:61]: Make it return `bool`, indicating whether it was done or failed within the context deadline. Remove the goroutine. Rename to `WaitDoneOrFailedCtx()`.

File: pkg/errors/gatherer.go [hunk 2]
```diff
@@ -44,6 +47,32 @@ func (g *Gatherer) Failed() <-chan struct{} {
 	return g.failed
 }
 
+// DoneOrFailed closes the returned channel when either:
+// all routines returned successfully OR
+// any routine returned with an error.
+func (g *Gatherer) DoneOrFailed() <-chan struct{} {
```
Comment (RmbRT) [inline pkg/errors/gatherer.go:53]: Make it return nothing, call the new `WaitDoneOrFailed()` as suggested. Rename to `WaitDoneOrFailed()`.

File: pkg/errors/gatherer.go [hunk 3]
```diff
@@ -44,6 +47,32 @@ func (g *Gatherer) Failed() <-chan struct{} {
...
+		case <-g.failed:
+		case <-g.wg.WaitCh():
+		case <-ctx.Done():
+		}
+		close(ret)
+	}()
+
+	return ret
+}
+
```
Comment (RmbRT) [inline pkg/errors/gatherer.go:75]: Add `Done() <-chan struct{}` returning the WaitGroup's `WaitCh()`.

File: pkg/errors/gatherer.go [hunk 4]
```diff
@@ -44,6 +47,32 @@ func (g *Gatherer) Failed() <-chan struct{} {
...
+		select {
+		case <-g.failed:
+		case <-g.wg.WaitCh():
+		case <-ctx.Done():
+		}
+		close(ret)
+	}()
+
+	return ret
+}
```
Comment (RmbRT) [inline pkg/errors/gatherer.go:74]: *(Cannot leave comment on unchanged lines)*
Also add a `WaitCtx(Context) bool` function that calls the waitgroup's `WaitCtx()`.

Comment (RmbRT) [reply pkg/errors/gatherer.go:74]: Might also return `error, bool`, but that's a bit clumsy IMO. Either way is fine with me.

**Review by RmbRT — CHANGES_REQUESTED** [review]

**Comment by manoranjith** [comment]
@ggwpez Where you able to figure out the reason for linter error ?
I am also getting the same error on my PRs as well. But when I try to run the linter locally, I do not get this error.

**Comment by matthiasgeihs** [comment]
> @ggwpez Where you able to figure out the reason for linter error ?
> I am also getting the same error on my PRs as well. But when I try to run the linter locally, I do not get this error.

It looks like the function that is used got deprecated.
https://staticcheck.io/docs/checks#SA1019
Would it be possible to rewrite that code without using the deprecated function?

**Comment by ggwpez** [comment]
> Update: I think we can just modify the panic message:
> `panic(fmt.Sprintf("Expected both interfaces to be zero or both non-zero, got %v, %v", left, right))`

Yes. Its just weird that i cannot locally reproduce the error, just as @manoranjith said.  
I will create a new PR for the fix.

File: pkg/errors/gatherer.go (hunk 1 above)
Comment (ggwpez) [reply pkg/errors/gatherer.go]: done

File: pkg/errors/gatherer.go (hunk 2 above)
Comment (ggwpez) [reply pkg/errors/gatherer.go]: done

File: pkg/errors/gatherer.go (hunk 3 above)
Comment (ggwpez) [reply pkg/errors/gatherer.go:70]: Would like to not blow up the MR, referenced in https://github.com/perun-network/poly-go/issues/5

File: pkg/errors/gatherer.go (hunk 4 above)
Comment (ggwpez) [reply pkg/errors/gatherer.go:69]: Would like to not blow up the MR, referenced in https://github.com/perun-network/poly-go/issues/5

**Review by RmbRT — APPROVED** [review]

**Pull request merged**